src/services/pathfinding.service.ts

```diff
@@ -1,54 +1,164 @@
   start: Location,
   end: Location
 ): Line[] => {
-  return [];
+  const { travelNodes } = buildingFloors.find(({ id }) => id === floorId);
+  const nodes = _.cloneDeep(travelNodes);
+
+  const edges: Line[] = traverseNodes(nodes);
+  const startEdge: Line = findClosestLine(edges, start);
+  const endEdge: Line = findClosestLine(edges, end);
+
+  const initial: TravelNode = {
+    id: nodes.length,
+    location: start,
+    children: [startEdge[0].id, startEdge[1].id]
+  };
+  const goal: TravelNode = {
+    id: nodes.length + 1,
+    location: end,
+    children: [endEdge[0].id, endEdge[1].id]
+  };
+
+  if (_.isEqual(startEdge, endEdge)) {
+    initial.children.push(goal.id);
+    goal.children.push(initial.id);
+  }
+  nodes.push(initial, goal);
+
+  nodes[startEdge[0].id].children.push(initial.id);
+  nodes[startEdge[0].id].children = nodes[startEdge[0].id].children.filter(
+    c => startEdge[1].id !== c
+  );
+
+  nodes[startEdge[1].id].children.push(initial.id);
+  nodes[startEdge[1].id].children = nodes[startEdge[1].id].children.filter(
+    c => startEdge[0].id !== c
+  );
+
+  nodes[endEdge[0].id].children.push(goal.id);
+  nodes[endEdge[0].id].children = nodes[endEdge[0].id].children.filter(
+    c => endEdge[1].id !== c
+  );
+
+  nodes[endEdge[1].id].children.push(goal.id);
+  nodes[endEdge[1].id].children = nodes[endEdge[1].id].children.filter(
+    c => endEdge[0].id !== c
+  );
+
+  const searchPath = search(nodes, initial, goal);
+
+  if (searchPath === null) return null;
+
+  const shortestPath: Line[] = [];
+  let { id } = goal;
+  while (id !== initial.id) {
+    const parentId: number = searchPath[id];
+    shortestPath.push([nodes[id], nodes[parentId]]);
+    id = parentId;
+  }
+  return shortestPath;
+};
+
+/**
+ * A* implementation for indoor pathfinding
+ * @param nodes
+ * @param initial
+ * @param goal
+ */
+const search = (
+  nodes: TravelNode[],
+  initial: TravelNode,
+  goal: TravelNode
+): {} => {
+  const closed = {};
+  const open: PriorityQueue<PQItem> = new PriorityQueue<PQItem>({
+    comparator: (a, b) => {
+      return f(nodes[a.id], goal, a.g) - f(nodes[b.id], goal, b.g);
+    }
+  });
+  open.queue({ id: initial.id, parent: null, g: 0 });
+  while (open.length > 0) {
+    const current = open.dequeue();
+    if (!(current.id in closed)) {
+      closed[current.id] = current.parent;
+      if (current.id === goal.id) return closed;
+      nodes[current.id].children.forEach(child => {
+        if (child in closed) return;
+        const g =
+          current.g +
+          getDistance(nodes[current.id].location, nodes[child].location);
+        open.queue({ id: child, parent: current.id, g });
+      });
+    }
+  }
+  return null;
+};
+
+/**
+ * The sum of the distance from the destination to a node and the distance from the start location to a node
+ * @param travelNode
+ * @param goal
+ * @param g
+ */
+const f = (travelNode: TravelNode, goal: TravelNode, g: number): number => {
```

Comment:
Maybe a more descriptive function name?

Reply:
![image](https://user-images.githubusercontent.com/36641869/77602595-40fad280-6ee4-11ea-848f-184ae9f59042.png)
We used the formula described above, and we figured that it is less confusing for us if we call this function f. I can add this formula as a comment, or change the function name to another name("aStarFunc" for example). What do you think?

Reply:
Yup, a small mention of that would be nice

Reply:
Is it ok to mention the formula briefly and put a link to the Wikipedia?